packages/ra-core/src/dataProvider/useDataProviderWithDeclarativeSideEffects.ts

```diff
@@ -0,0 +1,73 @@
+import {
+    useNotify,
+    useRedirect,
+    useRefresh,
+    useUnselectAll,
+} from '../sideEffect';
+import useDataProvider, { DataProviderHookFunction } from './useDataProvider';
+import { useCallback } from 'react';
+
+const useDataProviderWithDeclarativeSideEffects = (): DataProviderHookFunction => {
+    const dataProvider = useDataProvider();
+    const notify = useNotify();
+    const redirect = useRedirect();
+    const refresh = useRefresh();
+    const unselectAll = useUnselectAll();
+
+    return useCallback(
+        (type: string, resource: string, params: any, options: any = {}) => {
+            const convertToFunctionSideEffect = sideEffects => {
```

Comment:
I think you can move this method at the top of the file or in another file, and write unit tests.

Reply:
No he can't, because the scope of the function must contain the side effects grabbed from hooks just above

Reply:
It's not possible to pass them as params?